### ISSUE TITLE
Auto documentation?

### ISSUE THREAD

**Issue by rafaqz** [mauro3/Parameters.jl#59]
I've been wondering about documenting all of my default values so I thought I would share this idea here. I'm using DocStringExtensions.jl and Documenter.jl to build a lot of my documentation automatically.

It would be great if we could somehow get the default values into the struct documentation. DocStringExtensions.jl has $(FIELDS) hook that can add fields to one or all struct doc-strings. Somehow, that could include Parameters.jl default values too.

**Comment by mauro3** [comment]
Yes, sounds good. I don't know whether it would be needed in the Docs-packages or in Parameters.

PR welcome.

**Comment by ChrisRackauckas** [comment]
I think the idea is to have Parameters.jl automatically add docstrings, or extend docstrings. It would be cool if all of the kwargs get a nice documentation format, like:

```julia
@with_kw """
A representation of a chill number
""" MyStruct
  "The absolute value"
  a::Float64 = 1.0
  "The cool number"
  cool::Int = 2
end
```

then corresponds to:

```julia
"""
MyStruct([a,cool])

A representation of a chill number

Keyword arguments:
a::Float64 - The absolute value. Default value: 1.0
cool::Int - The cool number. Default value: 2
"""
...
```

**Comment by rafaqz** [comment]
Yep that seems perfect, just include the docstring in the macro. No deps either.

So we just need a new method of the macro:
```
macro kw_args(docstring, typedef)
```
and maybe some refactoring of `function with_kw()`? I'll have a go when I get time, but its probably less work than actually writing the docs...

**Comment by mauro3** [comment]
Well, I'm a bit hesitant to add such a large set of extra functionality (parsing and constructing doc-strings) to Parameter.jl.  I feel there should be an API in DocStringExtensions to hook into.  Maybe the field-doc strings could be augmented with their defaults values by Parameter.jl and the passed off to DocStringExtensions.

**Comment by rafaqz** [comment]
I can't picture how that would actually work - how do we get the kw_args to DocStringExtensions from the inner keyword constructor? Is the full AST accessible?

**Comment by rafaqz** [comment]
It should only be a few extra lines now I'm thinking about it, I'll put together a pull-requrest. I'll do the same for my [MetaParameters](https://github.com/rafaqz/MetaParameters.jl) to standardise. 
I think DocStringExtensions wont need to be touched using field-doc strings, that works already.

**Comment by mauro3** [comment]
> It should only be a few extra lines...

Famous last words ;-)

Yes, PR would be cool!

**Comment by rafaqz** [comment]
Only 6!

But the inevitable complication is that there is an existing bug in docstring handling that becomes apparent when you put a docstring on everything....

This test (line 134 ish) hits it, you can trigger it by adding a field doc-string:
```
# with type parameters and supertype
@with_kw mutable struct MT4_1{T} <: MM1
    " break things now "
    r::Int=5
    a::T
end
```
It sends args to keep_only_typparas, which calls `.head` on the string... So I stripped the strings out, which get me a Symbol/Exp conversion error, and after that I'm lost...

That function with_kw is pretty long to debug behind a macro! It might be nice to break it up so it can be unit tested, but that is really stretching past my 6 lines ;)

**Comment by mauro3** [comment]
Thanks!  I'll try to look at it but I'm also traveling/busy over the next two weeks.

Yeah, I know, `with_kw` should be refactored, probably using MacroTools.  I just haven't found the time/willingness yet.

**Comment by mauro3** [comment]
The latest master should have it fixed.  But do not the Julia issue https://github.com/JuliaLang/julia/issues/27656
(And maybe also https://github.com/JuliaLang/julia/issues/27091)

**Comment by rafaqz** [comment]
This still breaks on line 137 of the tests... I think it needs a field docstring in the test

@with_kw mutable struct MT4_1{T} <: MM1
    " break things now "
    r::Int=5
    a::T
end

**Comment by mauro3** [comment]
I pushed another try!

**Comment by rafaqz** [comment]
Thanks! All good. Tests a passing. We have documentation.

**Comment by mauro3** [comment]
This was closed by #61.
